examples/ShowPlayer.cpp

```diff
@@ -85,68 +92,111 @@ int ShowPlayer::Playback(unsigned int iterations,
   return ola::EXIT_OK;
 }
 
-void ShowPlayer::SendNextFrame() {
-  DmxBuffer buffer;
-  unsigned int universe;
-  ShowLoader::State state = m_loader.NextFrame(&universe, &buffer);
-  switch (state) {
-    case ShowLoader::END_OF_FILE:
-      HandleEndOfFile();
-      return;
-    case ShowLoader::INVALID_LINE:
-      m_client.GetSelectServer()->Terminate();
-      return;
-    default:
-      {}
-  }
 
-  state = RegisterNextTimeout();
+/**
+ * Start playback
+ */
+void ShowPlayer::Start() {
+  SeekTo(m_start);
+}
+
 
-  OLA_INFO << "Universe: " << universe << ": " << buffer.ToString();
-  ola::client::SendDMXArgs args;
-  m_client.GetClient()->SendDMX(universe, buffer, args);
-
-  switch (state) {
-    case ShowLoader::END_OF_FILE:
-      HandleEndOfFile();
-      return;
-    case ShowLoader::INVALID_LINE:
-      m_client.GetSelectServer()->Terminate();
-      return;
-    default:
-      {}
+/**
+ * Seek to @p seek_time in the show file
+ * @param seek_time the time (in milliseconds) to seek to
+ */
+void ShowPlayer::SeekTo(const unsigned int seek_time) {
+  // Seeking to a time before the playhead's position requires moving from the
+  // beginning of the file.  This could be optimized more if this happens
+  // frequently.
+  if (seek_time < m_playback_pos) {
+    m_loader.Reset();
+    m_playback_pos = 0;
   }
+
+  // Keep reading through the show file until desired time is reached.
+  ShowEntry entry;
+  unsigned int playhead_time = m_playback_pos;
+  do {
+    ShowLoader::State state = m_loader.NextEntry(&entry);
+    switch (state) {
+      case ShowLoader::END_OF_FILE:
+        HandleEndOfFile();
+        return;
+      case ShowLoader::INVALID_LINE:
+        m_client.GetSelectServer()->Terminate();
+        return;
+      default: {}
+    }
+    playhead_time += entry.next_wait;
+  } while (playhead_time < seek_time);
```

Comment:
Doesn't this want to be <=? Imagine the following file with a seek time of 5:
```
1 1,2,3
0
2 11,12,13
5
1 4,5,6
0
2 14,15,16
1
1 7,8,9
```

We want to output
`1 4,5,6`

As the first frame don't we? I think it will currently output 7,8,9?

Reply:
Stepping through with the debugger, I've got 11,12,13 immediately followed by 4,5,6 (0 ms delay).
I interpreted the requirements as "starting with the thing 5ms in," which would be 11,12,13.  Thinking about it now though, "start after 5ms has passed" (i.e. in this case 4,5,6) does seem more natural.  Will change.

Reply:
I wasn't expecting that result. I took it that the first line of the file is output at time zero, and then delay ms later the next line is output. I think 4,5,6 is the first data that is 5ms from the start (along with 14,15,16). i.e. the delay is attached to the following line (with an unlisted 0ms delay before the first line).

Remember in my slightly contrived (although fully supported) example, they are two universes, so 1,2,3 and 11,12,13 are happening at the same time.

Reply:
This looks to be what the old code did to me too:
Read frame
Register timeout to repeat
Output frame
Timeout expires, repeat process

Reply:
With some more testing, it seems there's a special case here.  If the start time lands at the end of a frame's timeout (as 5 does with your example) and the following frame has a timeout of 0 (as does 4,5,6), that following frame is skipped.  Enough to be noticeable?  Probably not, since timeout 0 means the next frame is broadcast near instantly, but still technically a problem.

Reply:
If they are two separate universes though, then that's potentially a bigger problem as one universe gets no data for 2 x timeperiods, which could also vary in length. With some protocols and few data changes, I think you could be talking 2+ seconds.

Reply:
This seems to be a bigger mess than I expected - it still follows the old process once playback has started, but finding the start point seems fraught with danger.

Reply:
I think the solution here will have to be:
- Track the value each universe is at while seeking
- Once the start point is found, send what the current state should be; this may be many lines ago for some universes.

Reply:
Alright, I think I've untangled this.
Using this file:
```
1 1,2,3
0
2 11,12,13
5
1 2,3,4
1
1 4,5,6
0
2 14,15,16
3
1 7,8,9
1
2 17,18,19
```
`--start 5` begins with `1: 2,3,4` `2: 11,12,13`
`--start 6` begins with `1: 4,5,6` `2: 11,12,13`

Reply:
> This seems to be a bigger mess than I expected - it still follows the old process once playback has started, but finding the start point seems fraught with danger.

Why do you think I didn't do the complicated bit? :laughing: 

> I think the solution here will have to be:
> 
>     * Track the value each universe is at while seeking
>     * Once the start point is found, send what the current state should be; this may be many lines ago for some universes.

Yeah, sounds reasonable, it potentially means we'd generate additional data, which we don't normally do, but I can't think of a better solution.



> Alright, I think I've untangled this.
> Using this file:

Would you agree your file maps to this table?

Time | Universe 1 | Universe 2
------------ | ------------- |  ------------- 
0 | 1,2,3 | 11,12,13
5 | 2,3,4 |
6 | 4,5,6 | 14,15,16
9 | 7,8,9 |
10 | | 17,18,19

> `--start 5` begins with `1: 2,3,4` `2: 11,12,13`
So that one looks fine.
> `--start 6` begins with `1: 4,5,6` `2: 11,12,13`
Is that going to output 14,15,16 immediately afterwards on universe 2?

The perfect solution when seeking is probably to look ahead and find the last point in time where you're still in the start time asked for (updating the stored values as you go), then output those values, then go into normal mode. Although I appreciate that's probably even more complicated. Both Art-Net and E1.31 (sACN) have a sync packet, so this would possibly help in making those synchronous.

Reply:
That table looks correct to me, and seems to match output when using `--start <time>`

`--start 6` gives:
```
examples/ShowPlayer.cpp:217: Universe: 2: 11,12,13
examples/ShowPlayer.cpp:217: Universe: 1: 4,5,6
examples/ShowPlayer.cpp:206: Registering timeout for 0ms
examples/ShowPlayer.cpp:217: Universe: 2: 14,15,16
examples/ShowPlayer.cpp:206: Registering timeout for 3ms
examples/ShowPlayer.cpp:217: Universe: 1: 7,8,9
examples/ShowPlayer.cpp:206: Registering timeout for 1ms
examples/ShowPlayer.cpp:217: Universe: 2: 17,18,19
```

Do you think it's worth adding another special case such that it would skip outputting 11,12,13 on U2 and send 14,15,16 instead?  I would have it not complete seeking (and therefore not output) until it encounters a non-0 timeout.

Reply:
Avoid that minor flaw turned out to be simpler than I expected.